Fix #17531: re-do build/dependency installation when build files not available

CaryXiao1 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #17531.
2. This PR does the following: Directly before downloading the build artifact in the `e2e_tests` job for Github Workflows, this PR makes the job check if the artifact created by the `build` job still exists. If it does not, it re-runs the commands from `build` to re-create and re-upload the build artifact before attempting to download.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
No proof of changes needed because it only modifies the E2E github workflow tests. Instead, I need to ensure that my code is able to correctly run tests **and is able to re-generate the build artifact if it has been deleted 24 hours after the build artifact was created**.

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

CaryXiao1 committed:
re-build artifact if actions/download-artifact fails

CaryXiao1 committed:
add newline at bottom of file

CaryXiao1 committed:
Merge branch 'develop' into rebuild-artifact

CaryXiao1 commented:
@U8NWXD are you able to re-run the e2e tests in `End-to-End and Lighthouse CI performance tests` for me? I currently don't have the permissions and want to check if my code can re-create the artifact, since it's been more than 24 hours since the original test.

CaryXiao1 commented:
Are you able to re-run one/some of the e2e tests in `End-to-End and Lighthouse CI performance tests` for me? I currently don't have the permissions and want to check if my code can re-create the artifact, since it's been more than 24 hours since the original test @U8NWXD PTAL

seanlip commented:
@CaryXiao1 yup sure thing! I just picked https://github.com/oppia/oppia/actions/runs/4600515715/jobs/8128260917?pr=17891 at random and reran it.

CaryXiao1 committed:
Merge branch 'develop' into rebuild-artifact

CaryXiao1 committed:
remove download of artifact after build_files is re-constructed

CaryXiao1 committed:
Merge remote-tracking branch 'origin/rebuild-artifact' into rebuild-artifact

CaryXiao1 committed:
merge work with upstream

CaryXiao1 committed:
only cache third_party/static when download fails

U8NWXD commented:
Hi @CaryXiao1 is the re-build working? Please lmk when it's working and you're ready for a review. Converting to a draft for now since it looks like this PR is still being tested

CaryXiao1 commented:
@U8NWXD The re-run that you did actually failed, so I added some more commits yesterday. I actually was planning to ask you to try and re-run one of the e2e tests since build_files is expired again - could you please pick some e2e test to re-run again?

seanlip commented:
Hi @CaryXiao1! I just kicked off a re-run of `classroomPageFileUploadFeatures` at random, FYI.

CaryXiao1 committed:
Merge branch 'develop' into rebuild-artifact

CaryXiao1 commented:
Alright, 'classroomPageFileUploadFeatures' succeeded over 24 hours after the artifact was created, so I would assume that the code is ready for review. @U8NWXD PTAL

U8NWXD commented on .github/workflows/e2e_and lighthouse_performance_tests.yml:
```diff
@@ -134,12 +134,54 @@ jobs:
+        run: python -m scripts.install_chrome_for_ci
+      - name: (Artifact re-build) Install ffmpeg
+        if: steps.download_artifact_1.outcome == 'failure'
+        run: sudo apt install ffmpeg
```
ffmpeg gets installed system-wide, so I don't think it'll be included in the cache. I think it needs to be installed no matter what. We probably missed this before video recording is disabled by default

CaryXiao1 replied:
Alright, thanks for the heads-up! I updated my branch to always install ffmpeg before the artifact download is attempted:
![image](https://user-images.githubusercontent.com/88044513/231360842-ee8ae06c-99f7-466c-94b2-d042fd828ee5.png)

U8NWXD commented on .github/workflows/e2e_and lighthouse_performance_tests.yml:
```diff
@@ -156,10 +198,13 @@ jobs:
+        if: steps.download_artifact_1.outcome != 'failure'
         run: python -m scripts.install_chrome_for_ci
       - name: Install ffmpeg
+        if: steps.download_artifact_1.outcome != 'failure'
```
It looks like we install these no matter what. Is there a reason to separately have them running if the outcome is `failure` and if the outcome is not `failure`?

CaryXiao1 replied:
Thanks for the suggestion - I was mainly worried that there was some series of dependencies that might have messed up something. I updated my PR to install all dependencies before trying the artifact download:
![image](https://user-images.githubusercontent.com/88044513/231361559-755f9b11-8a74-4a7c-968c-725e6009ccb9.png)

CaryXiao1 committed:
resolved suggestions to PR

CaryXiao1 committed:
Merge branch 'rebuild-artifact' of https://github.com/CaryXiao1/oppia into rebuild-artifact

CaryXiao1 committed:
Merge branch 'develop' into rebuild-artifact

CaryXiao1 committed:
Merge remote-tracking branch 'upstream/develop' into rebuild-artifact

CaryXiao1 committed:
Merge branch 'rebuild-artifact' of https://github.com/CaryXiao1/oppia into rebuild-artifact

CaryXiao1 committed:
remove extra space

CaryXiao1 commented:
I updated my code to reflect your requested changes - @U8NWXD PTAL

CaryXiao1 committed:
force overwrite in unzip

CaryXiao1 committed:
Merge branch 'develop' into rebuild-artifact

U8NWXD commented on .github/workflows/e2e_and lighthouse_performance_tests.yml:
```diff
@@ -134,15 +134,54 @@ jobs:
         run: |
           ls -la
-          unzip build_files.zip
+          unzip -o build_files.zip
```
Hmm. I think this reveals that I made a mistake in my earlier review. Sorry about that! It looks like we need to run the `Install Oppia Dependencies` step at different times depending on whether we can use the cache. If we can use the cache, then we run it after downloading the cache. If we cannot use the cache, then we have to run it earlier so that we can build the cache. ffmpeg can still be installed before we deal with caching, and I think we can put the install Chrome step after all the caching logic (it has to be later since we need the Python installed during `Install Oppia Dependencies`). Does that make sense? If I made another mistake, please let me know!

With the current code, we'll do some unnecessary dependency installation, which could be handled by the cache, even when the cache is available.

CaryXiao1 replied:
I see, so we want to install dependencies after unzipping so that we don't re-install when we do the force-unzip. I'll update that in the code!

CaryXiao1 committed:
fix redundant dependency installation

CaryXiao1 committed:
resolve merge conflit

CaryXiao1 committed:
remove conditional on installing Oppia Dependencies

CaryXiao1 committed:
Merge branch 'develop' into rebuild-artifact

CaryXiao1 commented:
I updated the code so that I no longer need to force-unzip the build artifact. Also, I changed the code so that all the extra steps for re-creating the artifact happen after the unzip step, so that there are less `if` checks when installing the dependencies @U8NWXD PTAL

U8NWXD reviewed (approved):
nice!

CaryXiao1 commented:
@seanlip can you merge the PR please? Thanks so much!
